# Math Project 307
contributors  
Ben 
Levi
Naidu
Shahada


# 1.6

In [3]:
import numpy as np

A=np.array([[2,3]
        ,[2,1]]); A.shape=(2,2)
print(A)
#TODO run on other matrices

W=[]
W.append(A)
for i in range(1,6):
    Q,R=np.linalg.qr(W[i-1],'complete')
    W.append(R@Q)
print(np.round(W[1:6],2))
eigenvalues = np.diagonal(W[5])
print('Eigenvalues are:' ,np.round(eigenvalues,2))
## check rounding


B=np.array([[1,1]
        ,[2,1]]); B.shape=(2,2)

C=np.array([[1,0,-1]
        ,[1,2,1]
        ,[-4,0,1]]); C.shape=(3,3)

D=np.array([[1,1,-1]
        ,[0,2,0]
        ,[-2,4,2]]); D.shape=(3,3)


ModuleNotFoundError: No module named 'numpy'

# 1.7

In [2]:
from Q_1point6 import QRalgo
import numpy as np

A=np.array([[2,3]
        ,[-1,-2]]); A.shape=(2,2)
QRalgo(A,6)
print("Using numpy function:",np.linalg.eigvals(A))

QRalgo(A,30)
print("Using numpy function:",np.linalg.eigvals(A))
# Can see the algorithm never converges on any eigvenvalues,
# looking at det(A-lamdaI)
# We can easily see A is not symmetric because
print(A,"A does not equal A^T: ",A.T)
# Therefore the matrix does not have real eigenvalues->
# it then has complex eigenvalues and therefore the QR
# algorithm can never converge to the eigenvalues.

A:

[[[ 4.    0.  ]
  [ 1.   -1.  ]]

 [[ 3.94  1.24]
  [ 0.24 -0.94]]

 [[ 4.01 -0.94]
  [ 0.06 -1.01]]

 [[ 4.    1.02]
  [ 0.02 -1.  ]]

 [[ 4.   -1.  ]
  [ 0.   -1.  ]]]
Eigenvalues are: [ 4. -1.]
Using numpy function: [ 4. -1.]
B:

[[[ 2.2   1.4 ]
  [ 0.4  -0.2 ]]

 [[ 2.44 -0.92]
  [ 0.08 -0.44]]

 [[ 2.41  1.01]
  [ 0.01 -0.41]]

 [[ 2.42 -1.  ]
  [ 0.   -0.42]]

 [[ 2.41  1.  ]
  [ 0.   -0.41]]]
Eigenvalues are: [ 2.41 -0.41]
Using numpy function: [ 2.41421356 -0.41421356]
C:

[[[ 2.   -0.   -3.89]
  [-0.73  2.18 -0.31]
  [-0.69  0.17 -0.18]]

 [[ 3.27  0.13  2.44]
  [-0.18  1.98  1.64]
  [-0.4  -0.04 -1.25]]

 [[ 2.96  0.16 -2.86]
  [-0.33  1.95 -1.81]
  [-0.11 -0.02 -0.91]]

 [[ 3.07  0.3   2.49]
  [-0.14  1.96  2.09]
  [-0.04 -0.01 -1.03]]

 [[ 3.03  0.34 -2.45]
  [-0.12  1.96 -2.21]
  [-0.01 -0.   -0.99]]]
Eigenvalues are: [ 3.03  1.96 -0.99]
Using numpy function: [ 2.  3. -1.]
D:

[[[ 3.   -1.07 -3.11]
  [-0.    2.    2.68]
  [ 0.   -0.    0.  ]]

 [[ 3.   -1.07  3.11]
  [

# 1.8


 Proof: 

 $Av=\lambda v, B=A+\alpha*I,
 (A+\alpha*I)*v=Av+\alpha*I*v=\lambda*v+\alpha*v=(\lambda+\alpha)*v,$



Showing that $\lambda+ \alpha$ is
 an eigenvalue of B, therefore if lamda is an eigenvalue of A,
 then $\lambda + \alpha$ is an eigenvalue of $B=A+\alpha*I$.


In [ ]:
from Q_1point6 import QRalgo
import numpy as np
A=np.array([[2,3]
        ,[-1,-2]]); A.shape=(2,2)
B=A+((0.9)*np.identity(2))
#shift
print("A:",A)
print("B:",B)
Bfinal=QRalgo(B,6)
#run Algo
print(Bfinal)
Bshiftback=Bfinal-((0.9)*np.identity(2))
#shift back
print(Bshiftback)
print(np.round(Bshiftback,1))
eigenvalues=np.diagonal(Bshiftback)
print("Eigenvalues using shift:",np.round(eigenvalues,1))
print("Using numpy function:",np.linalg.eigvals(A))



# 1.9 Find QR Factorization of $A^{k+1}$

$$(Q_0Q_1...Q_{k-1})A_k = A(Q_0Q_1...Q_{k-1})$$

$Q_0 = Q$  and $R_0 = R$ theorfore A = QR = $Q_0R_0$  
by inuction hypo. k = 1 

$$Q_0Q_1A_1 = AQ_oQ_1, \hspace{5mm} (AB)^T=B^TA^T$$
$$Q_0A_1 = Q_0Q_1(Q_0R_0)^T, \hspace{5mm} A(BC) =(AB)C \hspace{3mm} A = QR$$
$$=Q_0Q_1R^T_0Q^T_0$$
$$=Q_0Q_1A_1^T$$

from 1.3 $A_k =Q_kR_k$ and $A_{k+1}= R_kQ_k$

$$=Q_0Q_1(A_1^T)^T = Q_0Q_1A_1$$

for  k $\ge$ 1

$$(Q_0Q_1...Q_{k})(R_k...R_1R_0) = A(Q_0Q_1...Q_{k-1})(R_{k-1}...R_1R_0)$$
$$(Q_0Q_1...Q_{k-1})Q_{1}R_1(R_{k-1}...R_1R_0) = $$

$$(Q_0Q_1...Q_{k-1})A_1(R_{k-1}...R_1R_0) = $$

$$ (Q_0Q_1...Q_{k-1})A_1(R_{k-1}...R_1R_0)= Q_0R_0(Q_0Q_1...Q_{k-1})(R_{k-1}...R_1R_0)$$
$$= Q_0(R_0(Q_0Q_1...Q_{k-1}))^T(R_{k-1}...R_1R_0)$$
$$= Q_0(Q_0Q_1...Q_{k-1})^TR_0^T(R_{k-1}...R_1R_0)$$
$$= (Q_0(Q_0Q_1...Q_{k-1})^T)^TR_0^T(R_{k-1}...R_1R_0)$$
$$= (Q_0Q_1...Q_{k-1})Q_0^TR_0^T(R_{k-1}...R_1R_0)$$
$$= (Q_0Q_1...Q_{k-1})R_0Q_0(R_{k-1}...R_1R_0)$$
$$= (Q_0Q_1...Q_{k-1})A_1(R_{k-1}...R_1R_0)$$

$R_0Q_0 = A_1$ from hypo

Deduce $(Q_0Q_1...Q_{k})(R_k...R_1R_0) = QR$ of $A^{k+1}$

from 1.3 $A^{k+1} = R_kQ_k$

$$(Q_0Q_1...Q_{k})(R_k...R_1R_0)$$
$$(Q_0Q_1...Q_{k-1})Q_{k}R_k(K_{k-1}...R_1R_0)$$
$$((Q_0Q_1...Q_{k-1})Q_{k})^TR_k(K_{k-1}...R_1R_0)$$
$$Q_k^T(Q_0Q_1...Q_{k-1})^TR_k(K_{k-1}...R_1R_0)$$
$$Q_k^T((Q_0Q_1...Q_{k-1})^TR_k)^T(K_{k-1}...R_1R_0)$$
$$Q_k^TR_k^T(Q_0Q_1...Q_{k-1})(K_{k-1}...R_1R_0)$$
$$(Q_k^TR_k^T)^T(Q_0Q_1...Q_{k-1})(K_{k-1}...R_1R_0)$$

$$R_kQ_k(Q_0Q_1...Q_{k-1})(K_{k-1}...R_1R_0)$$
$$A^{k+1}(A_k...A_1A_0)$$
because we know from 1.3 $A_k$ is similar to A therefore $A_0$ is similar to $A_1$ and so on. therefore $A^{k+1}$ exists withng $(Q_0Q_1...Q_{k})(R_k...R_1R_0)$  therefore $(Q_0Q_1...Q_{k})(R_k...R_1R_0)$ must be the $A^{k+1}$ iteration.






# 1.11 QR with Shifts

Let A be a upper Hessenburg reduction with a traiganol. let $u_k$ be a eignevalue estimate.

by induction hypo. k =1 

$$Q_kR_k = A_{k-1} -u_kI$$
$$A_k = R_kQ_k + u_kI$$

$Q_1R_1 = A_0 - u_1I ,\hspace{3mm}A_1 = R_1Q_1  + u_1I$

$(Q_1R_1)^{-1}Q_1R_1 = (Q_1R_1)^{-1}A_0 - u_1I$

$I = (Q_1R_1)^{-1}A_0 - u_1I$
$A_0^{-1}I = (Q_1R_1)^{-1}A_0A_0^{-1} - u_1I$  
$A_0^{-1} = (Q_1R_1)^{-1}- u_1I$

because (Q_1R_1)^{-1} is symmetric then we have A^T = A and because col A are orthognal because it started as a upper hessenburg. we hav A^{-1} = A^T

$A_0 = (Q_1R_1)^{T}- u_1I$  
$A_0 = R^T_1Q^T_1- u_1I$   
$A_0 = (A_1 + u_1I)^T -u_1I$  
$A_0 = A_1^T + u_1I^T -u_1I$  
$A_0 = A_1 + u_1I -u_1I$  
$A_0 = A_1$

for K >= 1

from QR in 1.9 you have the folowing $$(Q_0Q_1...Q_{k})(R_k...R_1R_0)$$ 

$$=(Q_0Q_1...Q_{k-1})Q_{k}R_k(K_{k-1}...R_1R_0)$$

$$=(Q_0Q_1...Q_{k-1})(A_k -u_kI)(K_{k-1}...R_1R_0)$$

$$=(A_k(Q_0Q_1...Q_{k-1}) - (Q_0Q_1...Q_{k-1})u_kI)(K_{k-1}...R_1R_0)$$

$$=A_k(Q_0Q_1...Q_{k-1})(K_{k-1}...R_1R_0) - (K_{k-1}...R_1R_0)(Q_0Q_1...Q_{k-1})u_kI$$

$$=A_k(Q_0Q_1...Q_{k-1})(K_{k-1}...R_1R_0) - (K_{k-1}...R_1R_0)(Q_0Q_1...Q_{k-1})u_kI$$

$$=A_k(Q_0Q_1...Q_{k-2})Q_{k-1}K_{k-1}(R_{k-2}...R_1R_0) - (K_{k-1}...R_1)R_0Q_0(Q_1...Q_{k-1})u_kI$$

$$=A_k(Q_0Q_1...Q_{k-2})(A_{k-1}-u_{k-1}I)(R_{k-2}...R_1R_0) - (K_{k-1}...R_1)(A_0-u_0I)(Q_1...Q_{k-1})u_kI $$

$$...$$
simplfies to: 
$$=A_k...A_1A_0 - A_0A_1...A_k + A_0A_1...A_k(u_0...u_k)I - A_0A_1...A_k(u_0...u_k)I$$

because $A_{k+1}$ is similar to A_k from 1.9 

$A_k...A_1A_0 => A_{k+1}$

$A_0A_1...A_k => A_k$

and $A_0A_1...A_k(u_0...u_k)I - A_0A_1...A_k(u_0...u_k)I = 0$

thus you get $A_{k+1} - A_k = A_1 = A_0$ from hypothisis,
thus $A_{k+1}$ will have the same eigenvalue of $A_k$ 

# 2.1 The OuterProduct of the SVD

## (A)

$$A = U\Sigma V^T = [u_1,...,u_m] 
\begin{bmatrix}
\sigma_1 & ... & 0 & | & 0 \cr
0 & . & 0 & | & 0\cr
0 & 0 &\sigma_r & | & 0\cr
-&-&-&-&0\cr
0&0&0&0&0
\end{bmatrix} \begin{bmatrix} v_1^T\cr v^T_n \end{bmatrix}$$
wer ethe block matrix is D and R <= n we can partition U and vector V such that

$$ = [u_1...u_r|u_{r+1}...u_m] \begin{bmatrix} D &0 \cr 0 & 0 \end{bmatrix} \begin{bmatrix} v^T_1 \cr ... \cr v^T_r \cr - \cr v^T_{r+1} \cr ...\cr v^T_n \end{bmatrix}$$

using the definitation of matreix multiplication we can separate into two distinct forms because the D rxr matrix + 0mxm can be split as such. 

$$ = [u_1...u_r] D\begin{bmatrix} v^T_1 \cr ... \cr v^T_r \end{bmatrix} + [u_{r+1}...u_m] [0] \begin{bmatrix}v^T_{r+1} \cr ...\cr v^T_n \end{bmatrix} $$

$$ [u_{r+1}...u_m] [0] \begin{bmatrix}v^T_{r+1} \cr ...\cr v^T_n \end{bmatrix}  => 0$$

becuase the 0 matrix
thus you are left with 

$$ = [u_1...u_r] D\begin{bmatrix} v^T_1 \cr ... \cr v^T_r \end{bmatrix}$$

because D isa daignnol matrix containing all the sigmas from 1 to r by defintion of matrix mult you get 

$$= [\sigma_1u_1 + ...+\sigma_ru_r] \begin{bmatrix}v^T_{1} \cr ...\cr v^T_r \end{bmatrix} $$

$$= \sigma_1u_1v^T_1 + ...+\sigma_ru_rv^T_r $$

## (B)

$$A = \begin{bmatrix} -2 & 2\cr -1 & 1 \cr 2 & -2 \end{bmatrix} $$

$$A^TA =\begin{bmatrix} -2 & -1 & 2 &\cr 2 & 1 & -2 \end{bmatrix} \begin{bmatrix} -2 & 2\cr -1 & 1 \cr 2 & -2 \end{bmatrix}  = \begin{bmatrix} 9 & -9 \cr -9 & 9 \end{bmatrix}$$

$Det(A^TA -\lambda I) = 0$

$$=Det(\begin{bmatrix} 9-\lambda & -9\cr -9 & 9-\lambda \end{bmatrix})$$
$$= (9-\lambda)(9-\lambda) - (-9)^2$$
$$= \lambda^2 -18\lambda$$
$$=\lambda(\lambda - 18)$$

$\lambda_1 = 18$ and $\lambda_2 =0$

because we know from (A) that we only need u from 1 to r we only ever needto use $u_i = \frac{1}{\sigma_i}Av_i$ because anything above r is 0. for a reduce singluar expansion.

$\Sigma = \begin{bmatrix} 18 & 0 \cr 0 & 0\end{bmatrix}$

for $\lambda_1 =  18$

$$=\begin{bmatrix} 9-18 & -9\cr -9 & 9-18 \end{bmatrix} = \begin{bmatrix} -9 & -9 & |0\cr-9&-9& |0\end{bmatrix} = \begin{bmatrix} -9 & -9 \cr 0&0\end{bmatrix}$$

$-9x_1 -9x_2 = 0$  
$x_1 = -x_2$  
$x_2 = t$  

$x = [-x_2, x_2]$
x = t[-1,1]

$v_1 = \frac{1}{||x_1||}x_1 = \begin{bmatrix} -1/\sqrt{2} \cr 1/\sqrt{2} \end{bmatrix} $  


for $\lambda_2 = 0$
$$=\begin{bmatrix} 9 & -9\cr -9 & 9 \end{bmatrix} = \begin{bmatrix} 9 & -9 & |0\cr-9&9& |0\end{bmatrix} = \begin{bmatrix} 9 & -9 \cr 0&0\end{bmatrix}$$

$9x_1 -9x_2 = 0$  
$x_1 = x_2$  
$x_2 = t$  

$x = [x_2, x_2]$
x = t[1,1]

$v_1 = \frac{1}{||x_1||}x_1 = \begin{bmatrix} 1/\sqrt{2} \cr 1/\sqrt{2} \end{bmatrix} $ 

 

$V = \begin{bmatrix} -1/\sqrt{2}  & 1/\sqrt{2}\cr 1/\sqrt{2} & 1\sqrt{2} \end{bmatrix}$

$u_1 = \frac{1}{\sigma_1}Av_1 = \frac{1}{\sqrt{18}} \begin{bmatrix} -2 & 2\cr -1 & 1 \cr 2 & -2 \end{bmatrix} \begin{bmatrix} -1/\sqrt{2} \cr 1/\sqrt{2} \end{bmatrix} = \begin{bmatrix} 4/\sqrt{36} \cr 2/\sqrt{36} \cr -4/\sqrt{36} \end{bmatrix} $

for $u_2$ simple need to find values for $u_2^Tu_1 = 0$  

$u_2 =\begin{bmatrix} 1\cr2\cr2 \end{bmatrix} $

$U = \begin{bmatrix} 4/\sqrt{36} & 1 \cr 2/\sqrt{36} & 2 \cr -4/\sqrt{36} & 2 \end{bmatrix}$

whcih completes the SVD requirment for the question and the reduced part is simply the following because we have a D block matrix of size 1.

$$\sigma_1u_1v^T_1 =A_{reduced} = \sqrt{18}\begin{bmatrix} 4/\sqrt{36} \cr 2/\sqrt{36} \cr -4/\sqrt{36} \end{bmatrix}\begin{bmatrix} -1/\sqrt{2} \cr 1/\sqrt{2} \end{bmatrix}$$